src/shims/panic.rs

```diff
@@ -0,0 +1,223 @@
+        dest: PlaceTy<'tcx, Tag>,
+        ret: mir::BasicBlock,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        let tcx = &{this.tcx.tcx};
+
+        // fn __rust_maybe_catch_panic(
+        //     f: fn(*mut u8),
+        //     data: *mut u8,
+        //     data_ptr: *mut usize,
+        //     vtable_ptr: *mut usize,
+        // ) -> u32
+        let f = this.read_scalar(args[0])?.not_undef()?;
+        let data = this.read_scalar(args[1])?.not_undef()?;
+        let data_ptr = this.deref_operand(args[2])?;
+        let vtable_ptr = this.deref_operand(args[3])?;
+        let f_instance = this.memory.get_fn(f)?.as_instance()?;
+        this.write_null(dest)?;
+        trace!("__rust_maybe_catch_panic: {:?}", f_instance);
+
+        // In unwind mode, we tag this frame with some extra data.
+        // this lets '__rust_start_panic' know that it should jump back
+        // to thsi frame if a panic occurs (in MiriEvalContext::stack_pop)
+        if this.tcx.tcx.sess.panic_strategy() == PanicStrategy::Unwind {
+            this.frame_mut().extra.catch_panic = Some(UnwindData {
+                data: data.to_ptr()?,
+                data_ptr,
+                vtable_ptr,
+                dest: dest.clone(),
+                ret,
+            })
+        }
+
+        // Now we make a function call.
+        // TODO: consider making this reusable? `InterpCx::step` does something similar
+        // for the TLS destructors, and of course `eval_main`.
+        let mir = this.load_mir(f_instance.def, None)?;
+        let ret_place =
+            MPlaceTy::dangling(this.layout_of(tcx.mk_unit())?, this).into();
+        this.push_stack_frame(
+            f_instance,
+            mir.span,
+            mir,
+            Some(ret_place),
+            // Directly return to caller.
+            StackPopCleanup::Goto { ret: Some(ret), unwind: None },
+        )?;
+        let mut args = this.frame().body.args_iter();
+
+        let arg_local = args
+            .next()
+            .expect("Argument to __rust_maybe_catch_panic does not take enough arguments.");
+        let arg_dest = this.local_place(arg_local)?;
+        this.write_scalar(data, arg_dest)?;
+
+        args.next().expect_none("__rust_maybe_catch_panic argument has more arguments than expected");
+
+        // We ourselves will return `0`, eventually (because we will not return if we paniced).
+        this.write_null(dest)?;
+
+        // Don't fall through, we do *not* want to `goto_block`!
+        return Ok(());
+    }
+
+    #[inline(always)]
+    fn handle_stack_pop(
+        &mut self,
+        extra: FrameData<'tcx>,
+        unwinding: bool
+    ) -> InterpResult<'tcx, StackPopInfo> {
+        let this = self.eval_context_mut();
+        let mut res = if extra.is_panic_start {
+            trace!("unwinding: found is_panic_start");
+            StackPopInfo::StartUnwinding
+        } else {
+            StackPopInfo::Normal
+        };
```

Comment:
This is kind of messy still, with the mutable `res` variable...

What about
```rust
let res = if extra.is_panic_start {
            trace!("unwinding: found is_panic_start");
            StackPopInfo::StartUnwinding
        } else if let Some(unwind_data) = this.frame_mut().extra.catch_panic.take() {
            // the entire catch-the-panic story
        } else {
            StackPopInfo::Normal
        };
        this.memory.extra.stacked_borrows.borrow_mut().end_call(extra.call_id);
        Ok(res)
```

Reply:
Or maybe the `catch_panic` branch should go first, not sure?

Why do we mark the frame *above* as `catch_panic`? If you change `handle_catch_panic` to add this stuff to the new frame it pushes, then here we can just use `extra`, right?

Reply:
>  If you change handle_catch_panic to add this stuff to the new frame it pushes

That would make `catch_panic` *two* frames above the target frame, not one.

The idea here is that when we see `catch_panic`, we know that we're about to return *into* whatever frame wants to catch the panic. Therefore, we want to stop unwinding when we exit the frame *above* the target frame, so that we are no longer unwinding when we are back in the target frame.

Reply:
Currently we stop unwinding when the `catch_panic` frame becomes the top frame; I am proposing to instead stop unwinding when the `catch_panic` frame gets popped. So yes, that moves it up 1 level, but we still do `StopUnwinding` at the same time.

But how's that *two* frames above whatever the "target frame" is? In `__rust_maybe_catch_panic` we push a single stack frame, the frame of the closure around which catching happens. Currently you are marking the frame of the *caller* of `__rust_maybe_catch_panic` as special, but that seems odd to me -- the special frame is the one of the closure called by `__rust_maybe_catch_panic`, because when that closure unwinds, we want to turn it into a normal return.
So I'd say in your current scheme you are setting `catch_panic` in the frame *below* the target frame.

Reply:
Ah, I misunderstood what you meant. I agree that that makes more sense.

Reply:
@RalfJung: Your proposed modification to `res` is incorrect - we still may want to execute the 'catch-the-panic' block if `extra.is_panic_start` is `true`.

I think having `res` be mutable is the best solution here.

Reply:
>  we still may want to execute the 'catch-the-panic' block if extra.is_panic_start is true.

When can that happen?

Then just swap the order of the branches:
```rust
let res = if let Some(unwind_data) = this.frame_mut().extra.catch_panic.take() {
  ...
} else if extra.is_panic_start {
  ...
} else {
  ...
};
```

Reply:
I'm saying that we could theoretically start and stop unwinding on the same frame - this code shouldn't have to care.

Keeping the catch code separate also allows us to run `trace!("Popping during unwind!");` ` for both the initial frame and subsequent frame, without duplicating code.

Reply:
We start unwinding in `miri_start_panic` and end it in `__rust_maybe_catch_panic`, so it can't ever be the same frame. We can just ´assert!` that.

Reply:
Or with what I proposed in https://github.com/rust-lang/miri/pull/693/files#r342296266, we don't even need `is_panic_start` any more, simplifying the logic here as well.

Reply:
I've eliminated `is_panic_start`